### PR TITLE
rpc: refactor WSClient subscriptions API

### LINKED ISSUE #2756: Refactor WSClient subscriptions logic
We need to make subscriptions-related API more specific, see https://github.com/nspcc-dev/neo-go/pull/2749#discussion_r999882172.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2756.

### CODE REVIEW

File: pkg/rpcclient/wsclient.go [hunk 1]
```diff
@@ -237,13 +375,57 @@ readloop:
...
 				}
 			}
-			ok := make(map[chan<- Notification]bool)
+			ok := make(map[interface{}]bool)
+			if event == neorpc.MissedEventID {
+				c.subscriptionsLock.Lock()
+				for _, rcvr := range c.subscriptions {
+					ch := rcvr.Receiver()
+					rcvr.Close()
+					if ch != nil && !ok[ch] {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:384]: @roman-khimov, I just can't make it prettier, maybe you have some thoughts about implementation improvement? We have to keep information about channels that were already closed in order not to close them one more time and not to send something else to them.

File: pkg/rpcclient/wsclient.go [hunk 2]
```diff
@@ -394,102 +585,172 @@ func (c *WSClient) performUnsubscription(id string) error {
...
+					Value: b,
+				}
+			case <-c.done:
+				break waitloop
+			}
+		}
+		for range ch {
+		}
+	}()
+	return c.ReceiveBlocks(flt, ch)
```
Comment (roman-khimov) [inline pkg/rpcclient/wsclient.go:620]: Why should we bother with type-specific channels internally for old APIs? The old behavior is to push _everything_ into `c.Notifications` and it's easy to keep it alongside new logic.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/rpcclient/actor/waiter_test.go [hunk 3]
```diff
@@ -20,20 +19,20 @@ type AwaitableRPCClient struct {
...
-func (c *AwaitableRPCClient) SubscribeForNewBlocksWithChan(primary *int, since *uint32, till *uint32, rcvrCh chan<- rpcclient.Notification) (string, error) {
+func (c *AwaitableRPCClient) ReceiveBlocks(flt *neorpc.BlockFilter, rcvr chan<- *block.Block) (string, error) {
 	c.chLock.Lock()
 	defer c.chLock.Unlock()
-	c.subBlockCh = rcvrCh
+	c.subBlockCh = rcvr
 	return "1", nil
 }
-func (c *AwaitableRPCClient) SubscribeForTransactionExecutionsWithChan(state *string, container *util.Uint256, rcvrCh chan<- rpcclient.Notification) (string, error) {
+func (c *AwaitableRPCClient) ReceiveTransactionExecutions(flt *neorpc.ExecutionFilter, rcvr chan<- *state.AppExecResult) (string, error) {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/actor/waiter_test.go]: Fixed.

File: pkg/neorpc/types.go [hunk 4]
```diff
@@ -72,29 +72,32 @@ type (
...
+	// filter.
 	NotificationFilter struct {
 		Contract *util.Uint160 `json:"contract,omitempty"`
 		Name     *string       `json:"name,omitempty"`
 	}
 	// ExecutionFilter is a wrapper structure used for transaction execution
 	// events. It allows to choose failing or successful transactions based
-	// on their VM state.
+	// on their VM state and/or to choose execution event with the specified
+	// container. nil value treated as missing filter.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/neorpc/types.go]: Fixed.

File: pkg/rpcclient/wsclient.go [hunk 5]
```diff
@@ -237,13 +384,57 @@ readloop:
...
+						default:
+							panic("unknown receiver type")
+						}
+						ok[ch] = true
+					}
+					rcvr.Reset()
+				}
+				// TODO: here we're removing receiver's channel from the receivers list
+				//  to avoid "send to the closed channel" panic, but the user is still
+				//  responsible for unsubscription, is it OK?
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: pkg/rpcclient/wsclient.go [hunk 6]
```diff
@@ -237,13 +384,57 @@ readloop:
 					break readloop
 				}
 			}
-			ok := make(map[chan<- Notification]bool)
+			ok := make(map[interface{}]bool)
+			if event == neorpc.MissedEventID {
+				c.subscriptionsLock.Lock()
+				for _, rcvr := range c.subscriptions {
+					ch := rcvr.Receiver()
+					if ch != nil && !ok[ch] {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: pkg/rpcclient/wsclient.go [hunk 7]
```diff
@@ -237,13 +384,57 @@ readloop:
 					break readloop
 				}
 			}
-			ok := make(map[chan<- Notification]bool)
+			ok := make(map[interface{}]bool)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: pkg/rpcclient/wsclient.go [hunk 8]
```diff
@@ -237,13 +384,57 @@ readloop:
...
 			for _, rcvr := range c.subscriptions {
-				ntf := Notification{Type: event, Value: val}
-				if (rpcevent.Matches(rcvr, ntf) || event == neorpc.MissedEventID /*missed event must be delivered to each receiver*/) && !ok[rcvr.ch] {
-					ok[rcvr.ch] = true // strictly one notification per channel
-					rcvr.ch <- ntf     // this will block other receivers
+				ch := rcvr.Receiver()
+				if ch == nil {
+					continue
+				}
+				if rpcevent.Matches(rcvr, ntf) && !ok[ch] {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: pkg/rpcclient/wsclient.go [hunk 9]
```diff
@@ -394,102 +585,172 @@ func (c *WSClient) performUnsubscription(id string) error {
...
 // SubscribeForNewBlocks adds subscription for new block events to this instance
 // of the client. It can be filtered by primary consensus node index and/or block
-// index allowing to receive blocks since the specified index only, nil value is
-// treated as missing filter.
+// index allowing to receive blocks since/till the specified index only, nil value
+// is treated as missing filter.
 //
-// Deprecated: please, use SubscribeForNewBlocksWithChan. This method will be removed in future versions.
+// Deprecated: please, use ReceiveBlocks. This method will be removed in future versions.
 func (c *WSClient) SubscribeForNewBlocks(primary *int, sinceIndex, tillIndex *uint32) (string, error) {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: pkg/rpcclient/wsclient.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/wsclient.go]: Fixed.

File: docs/notifications.md [hunk 10]
```diff
@@ -346,9 +346,13 @@ Example:
 
 ### `transaction_executed` notification
 
-It contains the same result as from `getapplicationlog` method in the first
-parameter and no other parameters. The only difference from `getapplicationlog` is
-that it always contains zero in the `contract` field.
+It contains single parameter with six fields inside: container hash (block's or transaction's
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply docs/notifications.md]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**